ci-operator/config/openshift/kube-state-metrics/master.json

```diff
@@ -0,0 +1,35 @@
+{
+  "tag_specification": {
+    "cluster": "https://api.ci.openshift.org",
+    "namespace": "openshift",
+    "name": "origin-v3.11",
+    "tag": "",
+    "tag_overrides": {}
+  },
+  "base_images": {
+    "base": {
+      "cluster": "https://api.ci.openshift.org",
+      "namespace": "openshift",
+      "name": "origin-v3.11",
+      "tag": "base"
+    }
+  },
+  "canonical_go_repository": "github.com/openshift/kube-state-metrics",
+
+  "images": [
+    {
+      "from": "base",
+      "to": "kube-state-metrics",
+      "inputs": {
+        "root": {"as": ["0"]}
```

Comment:
If you are using a multi-stage dockerfile, this would tell openshift to use "root" (the base image from up above) as the base image for the first stage ("0").

Do you have a multistage docker build yet?  I didn't see it in master, didn't know if there was a PR I was missing.

Reply:
If you aren't using a multi stage build, you would define a "binary_build_commands" stanza that built the binary that you will `COPY` into the dockerfile, then you would remove the "inputs" stanza.

Reply:
E.g. in the Prometheus Operator _upstream_ project (github.com/coreos/prometheus-operator) we would build the binary without docker and then copy the binary during the `docker build` process.

If I understood @brancz correctly, using a multistage build is not an option for the production image build process. In addition it seem to require a single `docker build` command to build the image, instead of a `go build` and `docker build`. Thereby we have overwritten the upstream Dockerfile with [this](https://github.com/openshift/prometheus-operator/blob/openshift-master/Dockerfile), where the binary is build inside the image without multiple stages.

For the above described, should we keep `"root": {"as": ["0"]}`?

Reply:
So that means for now (3.11) I just remove:
```json
   "inputs": {
        "root": {"as": ["0"]}
    }
```
in all the `master.json` files, right?

---

On a side note, does

> the "from" key describes what replaces from (which is origin-base already).

mean, we could use our upstream `Dockerfile` with a different base and have the CI pipeline replace it on the fly?

Reply:
Yes.

> mean, we could use our upstream Dockerfile with a different base and have the CI pipeline replace it on the fly?

Yes